### ISSUE TITLE
Project exclusion does not work with unsupported project types, e.g. sqlproj

### ISSUE THREAD

**Issue by mrmartan** [security-code-scan/security-code-scan#201]
**Environment:**
 - Version: 5.1.1
 - Branch:
   - [ ] vs2019 
   - [ ] vs2017 
   - [ ] vs2015 
 - Installation/Running method:
   - [ ] Visual Studio Extension 
   - [ ] NuGet package
   - [X] Standalone tool
     - [X] DotNet Core Tool from NuGet
     - [ ] security-scan4x.zip from GitHub Release section
 - Operating System:
   - [X] Windows
   - [X] Linux
   - [ ] Mac

**Describe the bug**
I am trying to exclude incompatible projects, but the exclusion feature does not work properly. It behaves the same on Windows and Linux. Also the Windows machine does contain the SSDT MSBuild targets.

**Repro**

```bash
/mnt/c/source/dotnet$ security-scan ./DotnetDemo.sln --excl-proj=**/database/**
```

```
╔═╗┌─┐┌─┐┬ ┬┬─┐┬┌┬┐┬ ┬  ╔═╗┌─┐┌┬┐┌─┐  ╔═╗┌─┐┌─┐┌┐┌
╚═╗├┤ │  │ │├┬┘│ │ └┬┘  ║  │ │ ││├┤   ╚═╗│  ├─┤│││
╚═╝└─┘└─┘└─┘┴└─┴ ┴  ┴   ╚═╝└─┘─┴┘└─┘  ╚═╝└─┘┴ ┴┘└┘

.NET tool by Jaroslav Lobačevski v5.1.1.0


Loading solution './DotnetDemo.sln'
Resolve         0:00.0495419    DotnetDemo.Api.csproj (net5.0)
Resolve         0:01.1180255    UnitTests.csproj (net5.0)
Resolve         0:00.0019947    FunctionalTests.csproj (net5.0)
Resolve         0:00.0180871    IntegrationTests.csproj (net5.0)
Cannot open project '/mnt/c/source/dotnet/database/dotnet-demo.sqlproj' because the file extension '.sqlproj' is not associated with a language.
Finished loading solution './DotnetDemo.sln'
```

**Comment by jmguazzo** [comment]
With the previous version 5.0.1, the error raised by msbuild "Cannot open project..." didn't stop the console app. 
I don't know whether the fix from commit https://github.com/security-code-scan/security-code-scan/commit/c0bd8556f87b11585302fb0391c79b8009d0cd8b could be moved after the analysis ?

**Comment by JarLob** [comment]
There is https://github.com/security-code-scan/security-code-scan/blob/84e7c54640794a5caf20c5d7ce14d78ce29bf11b/SecurityCodeScan.Tool/Program.cs#L186 if you want to ignore any MSBuild errors. In VS2019 branch, not in release yet though.

**Comment by jmguazzo** [comment]
Even though the exclusion doesn't work, the parameter ```ignore-msbuild-errors``` doesn't block the console which is fine for my needs.

**Comment by ashwinnirantar** [comment]
ignore-msbuild-errors flag is not working for me. Neither is exlusion of project files. What can an alternative?

**Comment by waldo2188** [comment]
I have the same problem.
the workaround is to create a project.sln that does not contain the project that cannot be loaded.

But a fix will be a realy good thing ;)

**Comment by ashwinnirantar** [comment]
> I have the same problem.
> the workaround is to create a project.sln that does not contain the project that cannot be loaded.
> 
> But a fix will be a realy good thing ;)

For now this fix is okay, but I am planning to use this as part of our CI/CD pipelines, don't think this fix will help much :/

**Comment by ashwinnirantar** [comment]
@waldo2188 Does "--ignore-msbuild-errors" work for you?

**Comment by waldo2188** [comment]
I use the *v5.1.1.0* installed with the command `dotnet tool install --global security-scan`.

And when I try to use the command line option `--ignore-msbuild-errors` like this : `security-scan.exe ProjectA.sln --ignore-msbuild-errors`, the tool fail with the error below 

```shell

Loading solution '--ignore-msbuild-errors'
Unhandled exception. System.IO.FileNotFoundException: File not found : 'C:\Projets\ProjectA\--ignore-msbuild-errors'
   at Microsoft.CodeAnalysis.MSBuild.DiagnosticReporter.Report(DiagnosticReportingMode mode, String message, Func`2 createException)
   at Microsoft.CodeAnalysis.MSBuild.PathResolver.TryGetAbsoluteSolutionPath(String path, String baseDirectory, DiagnosticReportingMode reportingMode, String& absolutePath)
   at Microsoft.CodeAnalysis.MSBuild.MSBuildProjectLoader.LoadSolutionInfoAsync(String solutionFilePath, IProgress`1 progress, ILogger msbuildLogger, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.MSBuild.MSBuildWorkspace.OpenSolutionAsync(String solutionFilePath, ILogger msbuildLogger, IProgress`1 progress, CancellationToken cancellationToken)
   at SecurityCodeScan.Tool.Program.Main(String[] args) in C:\Projects\SCS\security-code-scan\SecurityCodeScan.Tool\Program.cs:line 296
   at SecurityCodeScan.Tool.Program.<Main>(String[] args)
```

**Comment by ashwinnirantar** [comment]
> I use the _v5.1.1.0_ installed with the command `dotnet tool install --global security-scan`.
> 
> And when I try to use the command line option `--ignore-msbuild-errors` like this : `security-scan.exe ProjectA.sln --ignore-msbuild-errors`, the tool fail with the error below
> 
> ```shell
> Loading solution '--ignore-msbuild-errors'
> Unhandled exception. System.IO.FileNotFoundException: File not found : 'C:\Projets\ProjectA\--ignore-msbuild-errors'
>    at Microsoft.CodeAnalysis.MSBuild.DiagnosticReporter.Report(DiagnosticReportingMode mode, String message, Func`2 createException)
>    at Microsoft.CodeAnalysis.MSBuild.PathResolver.TryGetAbsoluteSolutionPath(String path, String baseDirectory, DiagnosticReportingMode reportingMode, String& absolutePath)
>    at Microsoft.CodeAnalysis.MSBuild.MSBuildProjectLoader.LoadSolutionInfoAsync(String solutionFilePath, IProgress`1 progress, ILogger msbuildLogger, CancellationToken cancellationToken)
>    at Microsoft.CodeAnalysis.MSBuild.MSBuildWorkspace.OpenSolutionAsync(String solutionFilePath, ILogger msbuildLogger, IProgress`1 progress, CancellationToken cancellationToken)
>    at SecurityCodeScan.Tool.Program.Main(String[] args) in C:\Projects\SCS\security-code-scan\SecurityCodeScan.Tool\Program.cs:line 296
>    at SecurityCodeScan.Tool.Program.<Main>(String[] args)
> ```

I get the same issue, but if you modify the order of command line arguments the tool will run without this error. 

Try this `security-scan.exe --ignore-msbuild-errors ProjectA.sln`

**Comment by waldo2188** [comment]
@ashwinnirantar, you'r right, but the code analysis isn't done either.

The output of the command : 
```shell
security-scan.exe --ignore-msbuild-errors ProjectA.API.sln

╔═╗┌─┐┌─┐┬ ┬┬─┐┬┌┬┐┬ ┬  ╔═╗┌─┐┌┬┐┌─┐  ╔═╗┌─┐┌─┐┌┐┌
╚═╗├┤ │  │ │├┬┘│ │ └┬┘  ║  │ │ ││├┤   ╚═╗│  ├─┤│││
╚═╝└─┘└─┘└─┘┴└─┴ ┴  ┴   ╚═╝└─┘─┴┘└─┘  ╚═╝└─┘┴ ┴┘└┘

.NET tool by Jaroslav Lobačevski v5.1.1.0


Loading solution 'ProjectA.API.sln'
Resolve         0:00.0115036    ProjectA.DTO.csproj (net5.0)
Resolve         0:00.0045850    ProjectA.DTO.csproj (net48)
Resolve         0:00.2234109    ProjectA.Data.csproj (net5.0)
Resolve         0:00.5417227    ProjectA.API.csproj (net5.0)
Resolve         0:00.0146980    ProjectA.API.Tests.csproj (net5.0)
Resolve         0:00.0114899    SharedLibrary.csproj (net5.0)
Resolve         0:00.1744640    ProjectA.API.TestsIntegration.csproj (net5.0)
Impossible d'ouvrir le projet 'C:\Projets\ProjectA.api\docker-compose.dcproj', car l'extension de fichier '.dcproj' n'est pas associée à un langage.
Resolve         0:00.0170902    ProjectA.Manager.csproj (net5.0)
Resolve         0:00.0173092    ProjectA.Manager.Tests.csproj (net5.0)
Resolve         0:00.0064620    ProjectA.DTO.Tests.csproj (net5.0)
Finished loading solution 'ProjectA.API.sln'

```

**Comment by ashwinnirantar** [comment]
> @ashwinnirantar, you'r right, but the code analysis isn't done either.
> 
> The output of the command :
> 
> ```shell
> security-scan.exe --ignore-msbuild-errors ProjectA.API.sln
> 
> ╔═╗┌─┐┌─┐┬ ┬┬─┐┬┌┬┐┬ ┬  ╔═╗┌─┐┌┬┐┌─┐  ╔═╗┌─┐┌─┐┌┐┌
> ╚═╗├┤ │  │ │├┬┘│ │ └┬┘  ║  │ │ ││├┤   ╚═╗│  ├─┤│││
> ╚═╝└─┘└─┘└─┘┴└─┴ ┴  ┴   ╚═╝└─┘─┴┘└─┘  ╚═╝└─┘┴ ┴┘└┘
> 
> .NET tool by Jaroslav Lobačevski v5.1.1.0
> 
> 
> Loading solution 'ProjectA.API.sln'
> Resolve         0:00.0115036    ProjectA.DTO.csproj (net5.0)
> Resolve         0:00.0045850    ProjectA.DTO.csproj (net48)
> Resolve         0:00.2234109    ProjectA.Data.csproj (net5.0)
> Resolve         0:00.5417227    ProjectA.API.csproj (net5.0)
> Resolve         0:00.0146980    ProjectA.API.Tests.csproj (net5.0)
> Resolve         0:00.0114899    SharedLibrary.csproj (net5.0)
> Resolve         0:00.1744640    ProjectA.API.TestsIntegration.csproj (net5.0)
> Impossible d'ouvrir le projet 'C:\Projets\ProjectA.api\docker-compose.dcproj', car l'extension de fichier '.dcproj' n'est pas associée à un langage.
> Resolve         0:00.0170902    ProjectA.Manager.csproj (net5.0)
> Resolve         0:00.0173092    ProjectA.Manager.Tests.csproj (net5.0)
> Resolve         0:00.0064620    ProjectA.DTO.Tests.csproj (net5.0)
> Finished loading solution 'ProjectA.API.sln'
> ```

That's correct. I can confirm that I have the same issue. Is this intended or are we missing something here? Does the flag work for anyone?

**Comment by waldo2188** [comment]
I have cloned the project and launch the analyse from the branch vs2019. The parameter “ignore-msbuild-errors” is well in the source code and the analyse work fine. 
Maybe just a publication of a new nugget from the vs2019 branch will resolve our issue.

**Comment by JarLob** [comment]
Same here, created a vulnerable C# project. Added a new `sqlproj` project to the solution. Ran with `--ignore-msbuild-errors`. Got analysis warnings for the C# project. It would be better UX experience to hide the loading error. However the API callback provides only text error message. As I see from screenshots parsing the text is not an option because it can be localized.

**Comment by mrmartan** [comment]
@JarLob `--ignore-msbuild-errors` nonwithstanding, this issue is about `--excl-proj` not working properly and though once the version from VS2019 branch is released I can workaround it, I would rather like the `--excl-proj` to work properly.

**Comment by JarLob** [comment]
> It would be better UX experience to hide the loading error. However the API callback provides only text error message. As I see from screenshots parsing the text is not an option because it can be localized.

**Comment by mrmartan** [comment]
Not to disagree with that, but

> `/mnt/c/source/dotnet$ security-scan ./DotnetDemo.sln --excl-proj=**/database/**`

I don't understand. The loading errors seem to me to come from well loading the projects. I would expect that when I want to explicitly exclude any such unloadable projects, they will be simply ignored.

Are you saying it is intentional to load even the excluded projects? Am I misunderstanding the exclusion?

**Comment by JarLob** [comment]
The message is shown in a [callback](https://github.com/security-code-scan/security-code-scan/blob/vs2019/SecurityCodeScan.Tool/Program.cs#L284) from [workspace.OpenSolutionAsync call](https://github.com/security-code-scan/security-code-scan/blob/vs2019/SecurityCodeScan.Tool/Program.cs#L292). `--ignore-msbuild-errors` makes the message just informational (the project fails to load anyway, blame msbuild workspace loader). There is no API to distinguish if the failure is due unsupported project extension or it is more severe. There is no option to pass `OpenSolutionAsync` which projects to skip. `--excl-proj` is just a custom argument to exclude projects like unit tests from successfully loaded projects.

If you find a way how to prevent `OpenSolutionAsync` from even touching a specific project I will happily accept a PR.
Meanwhile `--ignore-msbuild-errors` or a custom solution where the project is excluded (like https://github.com/security-code-scan/security-code-scan/blob/vs2019/LGTM.sln for scanning SCS with https://lgtm.com) sounds as a reasonable workaround.

**Comment by ashwinnirantar** [comment]
> I have cloned the project and launch the analyse from the branch vs2019. The parameter “ignore-msbuild-errors” is well in the source code and the analyse work fine.
> Maybe just a publication of a new nugget from the vs2019 branch will resolve our issue.

@waldo2188 How did you build the binary after cloning? Can you point me to the instructions on building from source?

**Comment by ashwinnirantar** [comment]
Hey @JarLob thank you so much for releasing the new version. The ignore msbuild errors flag is working smoothly 👍

**Comment by mrmartan** [comment]
@JarLob I see. The trouble is we are including this scanner from GitLab SAST templates and would prefer not to have to override them. Would there be a way to make this work based on environment context, SLN metadata or env vars. Like for example detecting what current SDK supports and ignore related errors without specifying the switch. I will think about it more but solutions that come to my mind right now are messy.

Maybe the easiest thing right now would be to accept `--ignore-msbuild-errors` from env var.

**Comment by waldo2188** [comment]
@ashwinnirantar I didn't build the binary, just used the debug launcher of Visual Studio.

**Comment by ashwinnirantar** [comment]
> @ashwinnirantar I didn't build the binary, just used the debug launcher of Visual Studio.

Got it! Jaroslav has released a new version with support for this anyway. So all good :)

**Comment by JarLob** [comment]
> @JarLob I see. The trouble is we are including this scanner from GitLab SAST templates and would prefer not to have to override them. Would there be a way to make this work based on environment context, SLN metadata or env vars. Like for example detecting what current SDK supports and ignore related errors without specifying the switch. I will think about it more but solutions that come to my mind right now are messy.
> 
> Maybe the easiest thing right now would be to accept `--ignore-msbuild-errors` from env var.

Do you mean this one https://gitlab.com/gitlab-org/security-products/analyzers/security-code-scan? They are stuck with the old version of SCS AFAIK.
Really interested in how you use the dotnet global tool. You should be able to control the command line parameters.

**Comment by mrmartan** [comment]
Yeah, we are able to control the job, but by overriding parts of it. And since they update at their own pace I consider overriding parts of the job the worst of both worlds. I would ether include the template and forget about it or implement my whole own job.

```yaml
include:
  - template: Security/SAST.gitlab-ci.yml

security-code-scan-sast:
  variables:
    DOTNET_CLI_TELEMETRY_OPTOUT: 1
    DOTNET_NOLOGO: 1
  before_script:
    - |
      for i in $(find . -name "*.sqlproj")
      do
        dotnet sln $(find *.sln) remove ${i}
      done
```

Since GitLab CI variables are passed down pipelines and jobs using them would be IMHO the best solution. So what I would love to be able to do is:

```yaml
include:
  - template: Security/SAST.gitlab-ci.yml

variables:
  SECURITY_CODE_SCAN_IGNORE_MSBUILD_ERRORS: "true"
```

and so on.

**Comment by nickwaringa** [comment]
We also could really use the Gitlab solution above. SCS is kind of useless for us as solution (without rolling our own) in the current Gitlab pipelines.

**Comment by JarLob** [comment]
@mrmartan What I don't understand if [security-scan](https://www.nuget.org/packages/security-scan) was reading the parameter from env (this is easy to implement) how would that work with GitLab. AFAIK they download and extract a SCS dll (which is an engine of security-scan tool and doesn't have this "ignore errors" concept) from [Visual Studio Extension](https://marketplace.visualstudio.com/items?itemName=JaroslavLobacevski.SecurityCodeScanVS2019) for their pipelines https://gitlab.com/gitlab-org/security-products/analyzers/security-code-scan/-/blob/master/Dockerfile#L32

@nickwaringa I think if something is not usable in GitLab pipelines the best approach is to create a request in https://gitlab.com/gitlab-org/security-products/analyzers/security-code-scan

**Comment by mrmartan** [comment]
I had no idea they do it like that. I guess I just assumed they use this tool. I would have believed they link to this repo as well.

What they do seem like what a person unfamiliar with dotnet would do.

I guess I rollout my own version by using this tool. The env variable support would be useful though.

**Comment by mrmartan** [comment]
So the GitLab repository above links to https://security-code-scan.github.io/ which links to here. I think one naturally assumes they would use the standalone CLI tool and not install a VS extension.

**Comment by nickwaringa** [comment]
It is true @JarLob on the vsix installs (agreed on filing an issues). I pulled their image and picked completely through the entire thing last night. The Docker image runs Mono, and regardless of updates forces a pinned version of using MSBuild 15 (unless SCS is doing this). I attempted all manner of dotnet reinstall, upgrade, rebuild and mono uninstall reinstall on their image last night to try to force an MSBuild 16 version on to no avail.

I *think* that the initial dotnet execution is underpinned with MSBuild. Dotnet compilation fails, it attempts to fail gracefully to MSbuild and Nuget directly and they fail.

(Pardon my weak understanding of the .Net build world if wrong)
